### ISSUE TITLE
why do we dot product with look back?

### ISSUE THREAD

**Issue by seungbochoi** [lucidrains/reformer-pytorch#93]
I know it might be really dumb... but hope I could get an insight or analogy of look back and its purpose right before the dot product!

Thanks to your help, I could now understand why we can get [16,256,64,64] after 4 round hashing with 16 batches, and using this bucket location to get sorted Qs and Ks. 

Let say we ignore the batch(!6)for now... and only consider the 256 CHUNKs[256, 64,64].

So we use look back to make K[256,64,64] -> [256,64,64] by
extracting the 256th one and put it into TOP part. 

and then, we concat this one with the original K, so 
torch.cat([256,64,64], [256,64,64], dim=1) = [256,128,64]. 

So I can see that now our original chunk(64 buckets) are concat with one step before's chunk

and THEN, we make dot product with Q....which was [256,64,64]... 

so Q dot K = [256,64,64] * [256,128,64] -> [256,64,128]

I have no clue why we do this... can someone or @lucidrains please give me some examples or analogies of this? 

Thanks in advance...

Edit:
I thought tat we would make dot product with pure Q and its K(Q/norm(Q)) ..

**Comment by lucidrains** [comment]
@seungbochoi Nope not at all! Happy to explain!

One of the assumptions of the paper is that the qk tokens will be evenly distributed to the number of buckets specified. However, some tokens belonging to one bucket invariably spill over to the next since this is still stochastic. (For example, notice how the blue and yellow tokens are not all in one bucket in the figure below)

<img width="507" alt="Screen Shot 2020-05-08 at 8 27 23 PM" src="https://user-images.githubusercontent.com/108653/81462937-c8399800-916a-11ea-887c-345482412f77.png">

To maximize attention between tokens that have been put into the same bucket, they decided the solution was to have one bucket attend to its own bucket and the one before, with an assumption made in the paper that tokens in one bucket rarely exceed 2 buckets in size.

<img width="597" alt="Screen Shot 2020-05-08 at 8 33 25 PM" src="https://user-images.githubusercontent.com/108653/81463013-32523d00-916b-11ea-840f-1113ee628239.png">

Hope this answers your question!

**Comment by seungbochoi** [comment]
Thanks a lot for the comment! 

But I am still vague on [64,128] * [128,64] part! 

So I guess, we expand the original K with the next chunk to prevent the case that same bucket(let say blue one in second chunk) miss to attend the 1st chunk's BLUES! so I can see why we expand this K from [64,64] -> [128,64] THANKS!


But, 

still, why does it make sense to dot product the Q([64,64]) and expanded K[128,64] ? 

Thanks a lot again for the quick response Phil!

**Comment by lucidrains** [comment]
@seungbochoi maybe you missed this line! https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L296 K = Q / norm(Q) like you assumed

**Comment by seungbochoi** [comment]
I saw this ! But this line does not chamge the dimension from 64,64 to 128,64! 

I know that we need to have dot production of q and k(q/norm(q))! 

But am little confused by having 
[64,64] dot [64,128]. (And now i understood why we doubled the K!) 

Dot product is distance concept in vector space  and when we make dot product of first row of Q and first column of K, we are calculating the distance or similarity of those two vectors! 

But what is the underlying meaning of Q and expanded K? It seems this attention score is mo longer attention score of one sentence but something else..!

**Comment by lucidrains** [comment]
The underlying meaning is still the same. When you are doing full attention, you care most about dot products of vectors that are close in distance. LSH is theoretically bringing together QK vectors that are close into the same bucket, and then the lookback trick helps you not to exclude dot product pairs of bucketed vectors that have spilled over the boundaries of the bucket.
